### PR TITLE
alternator: fix DeleteTable return values to match DynamoDB's

### PR DESCRIPTION
**Pull request by Jager227**
It seems that Scylla has more values returned by DeleteTable operation than DynamoDB. In this patch I added a table status check when generating output. If we delete the table, values KeySchema, AttributeDefinitions and CreationDateTime won't be returned. The test has also been modified to check that these attributes are not returned.

Fixes scylladb#14132

### CODE REVIEW

File: test/alternator/test_table.py [hunk 1]
```diff
@@ -624,29 +624,20 @@ def test_delete_table_description(dynamodb):
     # Seems currently DynamoDB doesn't return CreationDateTime on DeleteTable so we disabled this assertion temporarily.
     # Check when DynamoDB will support it later and re-enable the field's verification.
 
-    # assert 'CreationDateTime' in got
+    assert not 'CreationDateTime' in got
```
Comment (nyh) [inline test/alternator/test_table.py:627]: Please also edit the comment above this assertion to better reflect reality, or just delete the comment entirely.

File: test/alternator/test_table.py [hunk 2]
```diff
@@ -624,29 +624,20 @@ def test_delete_table_description(dynamodb):
...
-        'AttributeDefinitions': [
-                    { 'AttributeName': 'p', 'AttributeType': 'S' },
-                    { 'AttributeName': 'c', 'AttributeType': 'S' },
-        ]
-    }
-    assert got['KeySchema'] == expected_schema['KeySchema']
+    assert not 'KeySchema' in got 
     # The list of attribute definitions may be arbitrarily reordered
-    assert multiset(got['AttributeDefinitions']) == multiset(expected_schema['AttributeDefinitions'])
+    assert not 'AttributeDefinitions' in got
```
Comment (nyh) [inline test/alternator/test_table.py:636]: Again, please edit - or just remove - the comment above this test.

**Review by nyh — CHANGES_REQUESTED** [review]
Thanks, looks good - I just added a request to please update (or just delete) comments that are no longer relevant after the code change.
Can you please confirm that after this patch, `test/alternator/run test_table.py::test_delete_table_description` passes both with and without the `--aws` option (on DynamoDB and ScyllaDB respectively)? Thanks.

**Comment by Jager227** [comment]
> Thanks, looks good - I just added a request to please update (or just delete) comments that are no longer relevant after the code change. Can you please confirm that after this patch, `test/alternator/run test_table.py::test_delete_table_description` passes both with and without the `--aws` option (on DynamoDB and ScyllaDB respectively)? Thanks.

I deleted the unnecessary comments and grouped up "assert not" statements for readability. I tested this patch on ScyllaDB and it passes perfectly, but unfortunately i have to ability to test it on DynamoDB due to lack of credentials and inability to obtain them.

**Comment by nyh** [comment]
> I deleted the unnecessary comments and grouped up "assert not" statements for readability. I tested this patch on ScyllaDB and it passes perfectly, but unfortunately i have no ability to test it on DynamoDB due to lack of credentials and inability to obtain them.

Thanks. I'll test it on DynamoDB myself in a few minutes

By the way, I DynamoDB has a "free tier" program where you 25 GB, 25 RCU and 25 WCU (reads and writes per second), for free. It's way more than enough for tests.

**Comment by Jager227** [comment]
> > I deleted the unnecessary comments and grouped up "assert not" statements for readability. I tested this patch on ScyllaDB and it passes perfectly, but unfortunately i have no ability to test it on DynamoDB due to lack of credentials and inability to obtain them.
> 
> Thanks. I'll test it on DynamoDB myself in a few minutes
> 
> By the way, I DynamoDB has a "free tier" program where you 25 GB, 25 RCU and 25 WCU (reads and writes per second), for free. It's way more than enough for tests.

That's cool, but i can't sign up from Russia, unfortunately :(

**Comment by nyh** [comment]
@Jager227 I verified that the updated test passes on AWS DynamoDB:
```
$ test/alternator/run --aws test_table.py::test_delete_table_description
...
============================================================================== 1 passed in 22.63s ===============================================================================
```

Since (unless I'm misunderstanding?) this is your first code contribution to the ScyllaDB project, I would like to ask you to please sign our contributor agreement, at https://www.scylladb.com/open-source-nosql-database/contributor-agreement/, and report here when you have done this. Thanks! (CC @avikivity)

**Review by nyh — APPROVED** [review]

**Comment by Jager227** [comment]
> @Jager227 I verified that the updated test passes on AWS DynamoDB:
> 
> ```
> $ test/alternator/run --aws test_table.py::test_delete_table_description
> ...
> ============================================================================== 1 passed in 22.63s ===============================================================================
> ```
> 
> Since (unless I'm misunderstanding?) this is your first code contribution to the ScyllaDB project, I would like to ask you to please sign our contributor agreement, at https://www.scylladb.com/open-source-nosql-database/contributor-agreement/, and report here when you have done this. Thanks! (CC @avikivity)

I already did this in advance.

**Comment by nyh** [comment]
Thanks. I'm waiting for the CI tests to finish, and I'll merge your pull request.

**Comment by nyh** [comment]
@Jager227 the CI discovered that your patch seems to have broken, on Alternator, the tests `test_describe_table.py::test_describe_table_schema` and `test_gsi.py::test_gsi_5_describe_table_schema`. For example, the former:

```
__________________________ test_describe_table_schema __________________________

test_table = dynamodb.Table(name='alternator_Test_1697537067996')

    def test_describe_table_schema(test_table):
        got = test_table.meta.client.describe_table(TableName=test_table.name)['Table']
        expected = { # Copied from test_table()'s fixture
            'KeySchema': [ { 'AttributeName': 'p', 'KeyType': 'HASH' },
                        { 'AttributeName': 'c', 'KeyType': 'RANGE' }
            ],
            'AttributeDefinitions': [
                        { 'AttributeName': 'p', 'AttributeType': 'S' },
                        { 'AttributeName': 'c', 'AttributeType': 'S' },
            ]
        }
        assert got['KeySchema'] == expected['KeySchema']
        # The list of attribute definitions may be arbitrarily reordered
>       assert multiset(got['AttributeDefinitions']) == multiset(expected['AttributeDefinitions'])
E       AssertionError: assert Counter() == Counter({froz...', 'S')}): 1})
E         Right contains 2 more items:
E         {frozenset({('AttributeName', 'p'), ('AttributeType', 'S')}): 1,
E          frozenset({('AttributeName', 'c'), ('AttributeType', 'S')}): 1}
E         Use -v to get more diff

test_describe_table.py:44: AssertionError
```

It seems your patch also removed the `AttributeDefinitions` output from DescribeTable, not just DeleteTable.

**Comment by Jager227** [comment]
> test_gsi.py::test_gsi_5_describe_table_schema

For me both this tests fail on scylladb:master as well, but maybe it's on my side only. Can you check them on master, please?

**Comment by Jager227** [comment]
I think i figured out what the problem was. I checked test_describe_table.py::test_describe_table_schema, test_gsi.py::test_gsi_5_describe_table_schema and test_table.py::test_delete_table_description. All 3 passed correctly this time.

File: alternator/executor.cc [hunk 3]
```diff
@@ -507,17 +504,21 @@ static rjson::value fill_table_description(schema_ptr schema, table_status tbl_s
...
-    // AttributeDefinitions for all key columns:
-    rjson::value attribute_definitions = rjson::empty_array();
-    for (auto& type : key_attribute_types) {
-        rjson::value key = rjson::empty_object();
-        rjson::add(key, "AttributeName", rjson::from_string(type.first));
-        rjson::add(key, "AttributeType", rjson::from_string(type.second));
-        rjson::push_back(attribute_definitions, std::move(key));
+    if (tbl_status != table_status::deleting) {
+        // Add base table's KeySchema and collect types for AttributeDefinitions:
+        executor::describe_key_schema(table_description, *schema, key_attribute_types);
```
Comment (nyh) [inline alternator/executor.cc:509]: Please move the declaration
```C++
std::unordered_map<std::string,std::string> key_attribute_types;
```
which you left in its original place to the new location here. Unless I'm missing something, there is no reason why the definition needs to stay many lines up

File: alternator/executor.cc [hunk 4]
```diff
@@ -507,17 +504,21 @@ static rjson::value fill_table_description(schema_ptr schema, table_status tbl_s
...
+        rjson::add(table_description, "CreationDateTime", rjson::value(creation_date_seconds));
```
Comment (nyh) [inline alternator/executor.cc:510]: Please move the setting of CreationDataTime a couple of lines up, the first thing of this if. In the current placement, it's in the middle of the code that handles the schema, so it's confusing.

**Review by nyh — CHANGES_REQUESTED** [review]
Thanks, looks good but I just found the code order a bit confusing so have a suggestion on how to improve it.

**Comment by nyh** [comment]
> #### 🟢 CI State: SUCCESS
> 
> ✅ - Build ✅ - Sanity Tests ✅ - Unit Tests

Congratulations!

File: alternator/executor.cc (hunk 3 above)
Comment (Jager227) [reply alternator/executor.cc:509]: This declaration is being used earlier in the code on string 488 (`executor::describe_key_schema(view_entry, *vptr, key_attribute_types);` ), so i can't move it lower.

Comment (nyh) [reply alternator/executor.cc:510]: Oh, I missed that...
But this brings another question: It's strange that the code above unconditionally (also in delete) adds the views' information in key_attribute_types, which won't be used, but only for the base table we don't. The loop above doesn't have the new `tbl_status != table_status::deleting` check you added here. I'm worried that if we DeleteTable a table which has a GSI or LSI, maybe we'll unnecessarily add its schema to the result.

I think we're missing a test for DeleteTable that would check the result of a DeleteTable of a table with GSI - I think the DeleteTable-result test we have in test_table.py doesn't have a GSI in its table, and we should probably add one?

Comment (Jager227) [reply alternator/executor.cc:510]: i'll add the != deleting condition to the loop, and then try to create a DeleteTable test for GSI/LSI case

Comment (Jager227) [reply alternator/executor.cc:510]: I just want to clarify, where exactly to put this condition?

Comment (nyh) [reply alternator/executor.cc:510]: There's a `if (!t.views().empty()) {` containing a loop of all the views. Do we need this for DeleteTable? Does DeleteTable return Local/GlobalSecondaryIndexes or doesn't it? If it doesn't, this whole if can be skipped. If it does, perhaps just something inside it which prints the schemas can be skipped.

This is why I always try to write the test before the development - to see what I *really* need to code. I understand you're having trouble with AWS. If you want, I can write a test for this and you'll just need to make the test pass with this code. Shall I write this test?

Comment (nyh) [reply alternator/executor.cc:510]: I'm writing this test, should be ready within an hour :-)

Comment (Jager227) [reply alternator/executor.cc:510]: > I'm writing this test, should be ready within an hour :-)

OK :slightly_smiling_face:

Comment (nyh) [reply alternator/executor.cc:510]: My test (I'll include it below) confirms that the GlobalSecondaryIndexes and LocalSecondaryIndexes should also be missing for DeleteTable, so perhaps the entire if() above should just be conditional on the state not being "deleting". I'll post the test below in a minute.

**Comment by nyh** [comment]
@Jager227 I propose this test, which demonstrates that DeleteTable not only drops KeySchema and AttributeDefinitions, it also drops LocalSecondaryIndexes and GlobalSecondaryIndexes from the output that DescribeTable returns.

Can you please add this test to your pull request (e.g., at the end of test_table.py), and fix your code to pass it? Thanks!

```python
# Above in test_delete_table_description() we tested the correctness of the
# return value of DeleteTable, and in particular that it doesn't include all
# the fields available in DescribeTable. The table in that test did not have
# secondary indexes, so in the following test we check the case of a table
# which does have secondary indexes. We note that whereas DescribeTable
# returns a description of the indexes (in fields "GlobalSecondaryIndexes"
# and "LocalSecondaryIndexes"), DeleteTable doesn't.
def test_delete_table_description_with_si(dynamodb):
    table = create_test_table(dynamodb,
        KeySchema=[{'AttributeName': 'p', 'KeyType': 'HASH'},
                   {'AttributeName': 'c', 'KeyType': 'RANGE'}],
        AttributeDefinitions=[ { 'AttributeName': 'p', 'AttributeType': 'S' },
                               { 'AttributeName': 'c', 'AttributeType': 'S' },
                               { 'AttributeName': 'x', 'AttributeType': 'S' }],
        GlobalSecondaryIndexes=[
            {   'IndexName': 'hello',
                'KeySchema': [{ 'AttributeName': 'x', 'KeyType': 'HASH' }],
                'Projection': { 'ProjectionType': 'ALL' }
            }],
        LocalSecondaryIndexes=[
            {   'IndexName': 'hithere',
                'KeySchema': [{'AttributeName': 'p', 'KeyType': 'HASH'},
                              {'AttributeName': 'x', 'KeyType': 'RANGE'}],
                'Projection': { 'ProjectionType': 'ALL' }
            }]
    )
    # Check that DescribeTable returns the table's KeySchema,
    # AttributeDefinitions, GlobalSecondaryIndexes and LocalSecondaryIndexes,
    # but DeleteTable does *not* return those, even though it could.
    got_describe = table.meta.client.describe_table(TableName=table.name)['Table']
    got_delete = table.delete()['TableDescription']
    assert got_describe['TableName'] == table.name
    assert got_delete['TableName'] == table.name
    for i in ['KeySchema', 'AttributeDefinitions', 'GlobalSecondaryIndexes', 'LocalSecondaryIndexes']:
        assert i in got_describe
        assert not i in got_delete
```

**Review by nyh — CHANGES_REQUESTED** [review]
Thanks! I still have a few requests, but I think we're getting close to the finish line :-)

File: test/alternator/test_table.py [hunk 5]
```diff
@@ -664,3 +647,39 @@ def test_delete_table_description_missing_fields(dynamodb):
 
     assert 'TableSizeBytes' in got
     assert 'ItemCount' in got
+# Above in test_delete_table_description() we tested the correctness of the
```
[the remark this answers is not in the document]
Comment (Jager227) [reply test/alternator/test_table.py:650]: Yes! I noticed it only after the commit

**Comment by Jager227** [comment]
I hope i didn't forget anything this time :-) .

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
Thanks @Jager227, merged into the "next" branch - will appear on "master" when another round of tests successfully completes.

@benipeled, as you can see above the CI (which I had to manually run, because this is an outside contributor) passed, but the "checks" box at the bottom doesn't say the CI passed (it apparently doesn't know about it?), and also says that "AWS Amplify Console Web Preview".  Any idea why these two things happened? Thanks.

**Comment by benipeled** [comment]
> @benipeled, as you can see above the CI (which I had to manually run, because this is an outside contributor) passed, but the "checks" box at the bottom doesn't say the CI passed (it apparently doesn't know about it?), and also says that "AWS Amplify Console Web Preview". Any idea why these two things happened? Thanks.

@nyh the CI ran against 00e9fd4afa12ad61a0e7304f2e6148687f44cdf7 which is a non-existent merge commit - I'm not sure how it happened, might be a force-push that removed it after the CI trigger. the CI checks show the status of the PR HEAD

Regarding the "AWS Amplify Console Web Preview" - you can ignore it, its docs CI should be reimplemented one day...

**Comment by nyh** [comment]
> Regarding the "AWS Amplify Console Web Preview" - you can ignore it, its docs CI should be reimplemented one day...

Does it ever work? If it doesn't, maybe it's better to disable it until it is reimplemented?

**Pull request closed without merging**